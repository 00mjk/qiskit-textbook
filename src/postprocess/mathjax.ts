// @ts-ignore
import { mjpage } from 'mathjax-node-page';

export async function processMathjax(htmlContent: string): Promise<string> {
  htmlContent = htmlContent.replace(/\\\\/g, '\\');
  const result = await mjpageAsync(htmlContent, {
    format: ["MathML", "TeX", "AsciiMath"], // determines type of pre-processors to run
    output: 'svg', // global override for output option; 'svg', 'html' or 'mml'
    tex: {}, // configuration options for tex pre-processor, cf. lib/tex.js
    ascii: {}, // configuration options for ascii pre-processor, cf. lib/ascii.js
    singleDollars: false, // allow single-dollar delimiter for inline TeX
    fragment: true, // return body.innerHTML instead of full document
    cssInline: true,  // determines whether inline css should be added
    displayMessages: false, // determines whether Message.Set() calls are logged
    displayErrors: false, // determines whether error messages are shown on the console
    undefinedCharError: false, // determines whether unknown characters are saved in the error array
    extensions: '', // a convenience option to add MathJax extensions
    fontURL: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/fonts/HTML-CSS', // for webfont urls in the CSS for HTML output
    MathJax: {}, // options MathJax configuration, see https://docs.mathjax.org
  });
  return result;
}

function mjpageAsync(content: string, mjpageConfig: any, mjnodeConfig: any = null): Promise<string> {
  return new Promise((resolve) => {
    mjpage(content, mjpageConfig, mjnodeConfig, (result: string) => {
      resolve(result);
    })
  })
}
