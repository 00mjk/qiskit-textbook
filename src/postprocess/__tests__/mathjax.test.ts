import { processMathjax } from '../mathjax';

describe('mathjax', () => {
  it('should render matrix correctly', async () => {
    const html = `
<div class="math notranslate nohighlight">
\\[\\begin{split}H =\\\\frac{1}{\\\\sqrt{2}}\\\\begin{pmatrix} 1 &amp; 1 \\\\\\\\ 1 &amp; -1 \\\\end{pmatrix}
.\\end{split}\\]</div>
    `;

    const result = await processMathjax(html);
    expect(result).toMatchSnapshot();
  });
});
