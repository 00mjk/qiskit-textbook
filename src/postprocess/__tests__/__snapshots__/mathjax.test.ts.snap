// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mathjax should render matrix correctly 1`] = `
"<div class=\\"math notranslate nohighlight\\">
<span class=\\"mjpage mjpage__block\\"><svg xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" width=\\"21.51ex\\" height=\\"6.509ex\\" style=\\"vertical-align: -2.671ex;\\" viewBox=\\"0 -1652.5 9261.3 2802.6\\" role=\\"img\\" focusable=\\"false\\" xmlns=\\"http://www.w3.org/2000/svg\\" aria-labelledby=\\"MathJax-SVG-1-Title\\">
<title id=\\"MathJax-SVG-1-Title\\">\\\\begin{split}H =\\\\frac{1}{\\\\sqrt{2}}\\\\begin{pmatrix} 1 &amp; 1 \\\\\\\\ 1 &amp; -1 \\\\end{pmatrix}
.\\\\end{split}</title>
<defs aria-hidden=\\"true\\">
<path stroke-width=\\"1\\" id=\\"E1-MJMATHI-48\\" d=\\"M228 637Q194 637 192 641Q191 643 191 649Q191 673 202 682Q204 683 219 683Q260 681 355 681Q389 681 418 681T463 682T483 682Q499 682 499 672Q499 670 497 658Q492 641 487 638H485Q483 638 480 638T473 638T464 637T455 637Q416 636 405 634T387 623Q384 619 355 500Q348 474 340 442T328 395L324 380Q324 378 469 378H614L615 381Q615 384 646 504Q674 619 674 627T617 637Q594 637 587 639T580 648Q580 650 582 660Q586 677 588 679T604 682Q609 682 646 681T740 680Q802 680 835 681T871 682Q888 682 888 672Q888 645 876 638H874Q872 638 869 638T862 638T853 637T844 637Q805 636 794 634T776 623Q773 618 704 340T634 58Q634 51 638 51Q646 48 692 46H723Q729 38 729 37T726 19Q722 6 716 0H701Q664 2 567 2Q533 2 504 2T458 2T437 1Q420 1 420 10Q420 15 423 24Q428 43 433 45Q437 46 448 46H454Q481 46 514 49Q520 50 522 50T528 55T534 64T540 82T547 110T558 153Q565 181 569 198Q602 330 602 331T457 332H312L279 197Q245 63 245 58Q245 51 253 49T303 46H334Q340 38 340 37T337 19Q333 6 327 0H312Q275 2 178 2Q144 2 115 2T69 2T48 1Q31 1 31 10Q31 12 34 24Q39 43 44 45Q48 46 59 46H65Q92 46 125 49Q139 52 144 61Q147 65 216 339T285 628Q285 635 228 637Z\\"></path>
<path stroke-width=\\"1\\" id=\\"E1-MJMAIN-3D\\" d=\\"M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z\\"></path>
<path stroke-width=\\"1\\" id=\\"E1-MJMAIN-31\\" d=\\"M213 578L200 573Q186 568 160 563T102 556H83V602H102Q149 604 189 617T245 641T273 663Q275 666 285 666Q294 666 302 660V361L303 61Q310 54 315 52T339 48T401 46H427V0H416Q395 3 257 3Q121 3 100 0H88V46H114Q136 46 152 46T177 47T193 50T201 52T207 57T213 61V578Z\\"></path>
<path stroke-width=\\"1\\" id=\\"E1-MJMAIN-32\\" d=\\"M109 429Q82 429 66 447T50 491Q50 562 103 614T235 666Q326 666 387 610T449 465Q449 422 429 383T381 315T301 241Q265 210 201 149L142 93L218 92Q375 92 385 97Q392 99 409 186V189H449V186Q448 183 436 95T421 3V0H50V19V31Q50 38 56 46T86 81Q115 113 136 137Q145 147 170 174T204 211T233 244T261 278T284 308T305 340T320 369T333 401T340 431T343 464Q343 527 309 573T212 619Q179 619 154 602T119 569T109 550Q109 549 114 549Q132 549 151 535T170 489Q170 464 154 447T109 429Z\\"></path>
<path stroke-width=\\"1\\" id=\\"E1-MJMAIN-221A\\" d=\\"M95 178Q89 178 81 186T72 200T103 230T169 280T207 309Q209 311 212 311H213Q219 311 227 294T281 177Q300 134 312 108L397 -77Q398 -77 501 136T707 565T814 786Q820 800 834 800Q841 800 846 794T853 782V776L620 293L385 -193Q381 -200 366 -200Q357 -200 354 -197Q352 -195 256 15L160 225L144 214Q129 202 113 190T95 178Z\\"></path>
<path stroke-width=\\"1\\" id=\\"E1-MJMAIN-28\\" d=\\"M94 250Q94 319 104 381T127 488T164 576T202 643T244 695T277 729T302 750H315H319Q333 750 333 741Q333 738 316 720T275 667T226 581T184 443T167 250T184 58T225 -81T274 -167T316 -220T333 -241Q333 -250 318 -250H315H302L274 -226Q180 -141 137 -14T94 250Z\\"></path>
<path stroke-width=\\"1\\" id=\\"E1-MJMAIN-2212\\" d=\\"M84 237T84 250T98 270H679Q694 262 694 250T679 230H98Q84 237 84 250Z\\"></path>
<path stroke-width=\\"1\\" id=\\"E1-MJMAIN-29\\" d=\\"M60 749L64 750Q69 750 74 750H86L114 726Q208 641 251 514T294 250Q294 182 284 119T261 12T224 -76T186 -143T145 -194T113 -227T90 -246Q87 -249 86 -250H74Q66 -250 63 -250T58 -247T55 -238Q56 -237 66 -225Q221 -64 221 250T66 725Q56 737 55 738Q55 746 60 749Z\\"></path>
<path stroke-width=\\"1\\" id=\\"E1-MJSZ3-28\\" d=\\"M701 -940Q701 -943 695 -949H664Q662 -947 636 -922T591 -879T537 -818T475 -737T412 -636T350 -511T295 -362T250 -186T221 17T209 251Q209 962 573 1361Q596 1386 616 1405T649 1437T664 1450H695Q701 1444 701 1441Q701 1436 681 1415T629 1356T557 1261T476 1118T400 927T340 675T308 359Q306 321 306 250Q306 -139 400 -430T690 -924Q701 -936 701 -940Z\\"></path>
<path stroke-width=\\"1\\" id=\\"E1-MJSZ3-29\\" d=\\"M34 1438Q34 1446 37 1448T50 1450H56H71Q73 1448 99 1423T144 1380T198 1319T260 1238T323 1137T385 1013T440 864T485 688T514 485T526 251Q526 134 519 53Q472 -519 162 -860Q139 -885 119 -904T86 -936T71 -949H56Q43 -949 39 -947T34 -937Q88 -883 140 -813Q428 -430 428 251Q428 453 402 628T338 922T245 1146T145 1309T46 1425Q44 1427 42 1429T39 1433T36 1436L34 1438Z\\"></path>
<path stroke-width=\\"1\\" id=\\"E1-MJMAIN-2E\\" d=\\"M78 60Q78 84 95 102T138 120Q162 120 180 104T199 61Q199 36 182 18T139 0T96 17T78 60Z\\"></path>
</defs>
<g stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" transform=\\"matrix(1 0 0 -1 0 0)\\" aria-hidden=\\"true\\">
<g transform=\\"translate(167,0)\\">
<g transform=\\"translate(-11,0)\\">
<g transform=\\"translate(0,65)\\">
 <use xlink:href=\\"#E1-MJMATHI-48\\" x=\\"0\\" y=\\"0\\"></use>
 <use xlink:href=\\"#E1-MJMAIN-3D\\" x=\\"1166\\" y=\\"0\\"></use>
<g transform=\\"translate(1944,0)\\">
<g transform=\\"translate(397,0)\\">
<rect stroke=\\"none\\" width=\\"1454\\" height=\\"60\\" x=\\"0\\" y=\\"220\\"></rect>
 <use xlink:href=\\"#E1-MJMAIN-31\\" x=\\"476\\" y=\\"676\\"></use>
<g transform=\\"translate(60,-920)\\">
 <use xlink:href=\\"#E1-MJMAIN-221A\\" x=\\"0\\" y=\\"39\\"></use>
<rect stroke=\\"none\\" width=\\"500\\" height=\\"60\\" x=\\"833\\" y=\\"780\\"></rect>
 <use xlink:href=\\"#E1-MJMAIN-32\\" x=\\"833\\" y=\\"0\\"></use>
</g>
</g>
</g>
<g transform=\\"translate(3916,0)\\">
 <use xlink:href=\\"#E1-MJSZ3-28\\"></use>
<g transform=\\"translate(903,0)\\">
<g transform=\\"translate(-11,0)\\">
 <use xlink:href=\\"#E1-MJMAIN-31\\" x=\\"0\\" y=\\"650\\"></use>
 <use xlink:href=\\"#E1-MJMAIN-31\\" x=\\"0\\" y=\\"-750\\"></use>
</g>
<g transform=\\"translate(1490,0)\\">
 <use xlink:href=\\"#E1-MJMAIN-31\\" x=\\"389\\" y=\\"650\\"></use>
<g transform=\\"translate(0,-750)\\">
 <use xlink:href=\\"#E1-MJMAIN-2212\\" x=\\"0\\" y=\\"0\\"></use>
 <use xlink:href=\\"#E1-MJMAIN-31\\" x=\\"778\\" y=\\"0\\"></use>
</g>
</g>
</g>
 <use xlink:href=\\"#E1-MJSZ3-29\\" x=\\"3839\\" y=\\"-1\\"></use>
</g>
 <use xlink:href=\\"#E1-MJMAIN-2E\\" x=\\"8659\\" y=\\"0\\"></use>
</g>
</g>
</g>
</g>
</svg></span></div>
    "
`;
